legacy/builder/hardware_loader.go

```diff
@@ -39,8 +39,22 @@ func (s *HardwareLoader) Run(ctx *types.Context) error {
 				ctx.Info(tr("Error loading hardware platform: %[1]s", err.Error()))
 			}
 		}
-		ctx.PackageManager = pm
+
+		if !ctx.CanUseCachedTools {
+			if ctx.BuiltInToolsDirs != nil {
+				pmb.LoadToolsFromBundleDirectories(ctx.BuiltInToolsDirs)
+			}
+
+			ctx.CanUseCachedTools = true
+		}
+
+		pm := pmb.Build()
+		pme, _ /* never release... */ := pm.NewExplorer()
```

Comment:
this is really ugly, maybe bringing this inside the arduino builder should solve the problem

Reply:
Unfortunately it's more complicated than that, we have to keep it for now.
Let's open a dedicated issue to keep track of it and fix it in the future.